File: sdks/typescript/test/client.test.js
import assert from "node:assert/strict";
import { readdir, readFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import test from "node:test";
import { fileURLToPath } from "node:url";
import {
  CtxCliError,
  CtxParseError,
  CtxTimeoutError,
  CtxUnsupportedError,
  CtxValidationError,
  AGENT_HISTORY_V1_VERSION,
  createHostedAgentHistoryClient,
  createLocalAgentHistoryClient,
} from "../src/index.js";
import { runDogfoodToy } from "../examples/dogfood-toy.js";

const repoRoot = join(dirname(fileURLToPath(import.meta.url)), "..", "..", "..");

function mockClient(handler) {
  const calls = [];
  const client = createLocalAgentHistoryClient({
    dataRoot: "/tmp/ctx-sdk-test",
    runner: async (request) => {
      calls.push(request);
      return handler(request);
    },
  });
  return { client, calls };
}

test("wraps status, init, sources, import, and sync CLI commands", async () => {
  const { client, calls } = mockClient(({ args }) => ({
    stdout: JSON.stringify({ initialized: true, sources: [{ provider: "codex" }], args }),
  }));

  const status = await client.status();
  await client.init({ catalogOnly: true });
  const sources = await client.sources();
  const imported = await client.import({ provider: "codex", resume: true });
  await client.sync({ all: true });

  assert.equal(status.contractVersion, AGENT_HISTORY_V1_VERSION);
  assert.equal(status.operation, "status");
  assert.equal(status.status.initialized, true);
  assert.equal(sources.sources[0].provider, "codex");
  assert.equal(imported.operation, "import");

  assert.deepEqual(
    calls.map((call) => call.args),
    [
      ["--data-root", "/tmp/ctx-sdk-test", "status", "--json"],
      [
        "--data-root",
        "/tmp/ctx-sdk-test",
        "setup",
        "--json",
        "--progress",
        "none",
        "--catalog-only",
      ],
      ["--data-root", "/tmp/ctx-sdk-test", "sources", "--json"],
      [
        "--data-root",
        "/tmp/ctx-sdk-test",
        "import",
        "--json",
        "--progress",
        "none",
        "--provider",
        "codex",
        "--resume",
      ],
      [
        "--data-root",
        "/tmp/ctx-sdk-test",
        "import",
        "--json",
        "--progress",
        "none",
        "--all",
      ],
    ],
  );
});

test("builds search flags and normalizes nested CLI search output", async () => {
  const { client, calls } = mockClient(() =>
    JSON.stringify({
      query: "retry handling",
      generated_at: "2026-07-01T12:00:00Z",
      freshness: { mode: "off", status: "skipped", source_count: 1, totals: {} },
      retrieval: {
        requested_mode: "hybrid",
        effective_mode: "lexical",
        semantic_weight: 0.0,
        semantic_status: "fallback",
        semantic_fallback_code: "semantic_retrieval_failed",
        semantic_fallback: "semantic_retrieval_failed",
        coverage: {
          embedded_items: 4,
          embedded_chunks: 9,
          searchable_items: 12,
          indexed_now: 1,
        },
        diagnostics: { query_embed_ms: 2, vector_scan_ms: 3 },
      },
      results: [
        {
          ctx_event_id: "00000000-0000-0000-0000-000000000101",
          ctx_session_id: "00000000-0000-0000-0000-000000000102",
          provider_session_id: "codex-session",
          event_seq: 7,
          result_type: "event",
          result_scope: "event",
          source_path: "/tmp/session.jsonl",
          source_exists: true,
          why_matched: ["text"],
          citations: [
            {
              target_type: "event",
              ctx_event_id: "00000000-0000-0000-0000-000000000101",
              ctx_session_id: "00000000-0000-0000-0000-000000000102",
              source_path: "/tmp/session.jsonl",
              source_exists: true,
            },
          ],
        },
      ],
      pagination: { next_cursor: "page-2", has_more: true },
      truncation: { truncated: false },
    }),
  );

  const result = await client.search("retry handling", {
    terms: ["timeout", "backoff"],
    limit: 5,
    provider: "codex",
    workspace: "ctx",
    since: "30d",
    primaryOnly: true,
    eventType: "message",
    file: "crates/foo/src/lib.rs",
    session: "00000000-0000-0000-0000-000000000001",
    events: true,
    backend: "hybrid",
    semanticWeight: 0.8,
    refresh: "off",
    includeCurrentSession: true,
  });

  assert.equal(result.contractVersion, AGENT_HISTORY_V1_VERSION);
  assert.equal(result.operation, "search");
  assert.equal(result.search.generatedAt, "2026-07-01T12:00:00Z");
  assert.equal(result.search.freshness.sourceCount, 1);
  assert.equal(result.search.results[0].ctxEventId, "00000000-0000-0000-0000-000000000101");
  assert.equal(result.search.results[0].ctxSessionId, "00000000-0000-0000-0000-000000000102");
  assert.equal(result.search.results[0].providerSessionId, "codex-session");
  assert.equal(result.search.results[0].eventSeq, 7);
  assert.equal(result.search.results[0].resultType, "event");
  assert.equal(result.search.results[0].resultScope, "event");
  assert.equal(result.search.results[0].sourcePath, "/tmp/session.jsonl");
  assert.equal(result.search.results[0].sourceExists, true);
  assert.equal(result.search.results[0].whyMatched[0], "text");
  assert.equal(result.search.results[0].citations[0].targetType, "event");
  assert.equal(result.search.results[0].citations[0].sourcePath, "/tmp/session.jsonl");
  assert.equal(result.search.retrieval.requestedMode, "hybrid");
  assert.equal(result.search.retrieval.effectiveMode, "lexical");
  assert.equal(result.search.retrieval.semanticWeight, 0.0);
  assert.equal(result.search.retrieval.semanticFallbackCode, "semantic_retrieval_failed");
  assert.equal(result.search.retrieval.semanticFallback, "semantic_retrieval_failed");
  assert.equal(result.search.retrieval.coverage.embeddedItems, 4);
  assert.equal(result.search.retrieval.coverage.indexedNow, 1);
  assert.equal(result.search.retrieval.diagnostics.queryEmbedMs, 2);
  assert.equal(result.search.pagination.nextCursor, "page-2");
  assert.equal(result.search.pagination.hasMore, true);

  assert.deepEqual(calls[0].args, [
    "--data-root",
    "/tmp/ctx-sdk-test",
    "search",
    "retry handling",
    "--term",
    "timeout",
    "--term",
    "backoff",
    "--limit",
    "5",
    "--provider",
    "codex",
    "--workspace",
    "ctx",
    "--since",
    "30d",
    "--primary-only",
    "--event-type",
    "message",
    "--file",
    "crates/foo/src/lib.rs",
    "--session",
    "00000000-0000-0000-0000-000000000001",
    "--events",
    "--backend",
    "hybrid",
    "--semantic-weight",
    "0.8",
    "--refresh",
    "off",
    "--include-current-session",
    "--json",
  ]);
});

test("omits semantic search override flags when unset", async () => {
  const { client, calls } = mockClient(() => JSON.stringify({ query: "default", results: [] }));

  await client.search("default");

  assert.equal(calls[0].args.includes("--backend"), false);
  assert.equal(calls[0].args.includes("--semantic-weight"), false);
});

test("rejects search without query, term, or file before invoking CLI", async () => {
  const { client, calls } = mockClient(() => {
    throw new Error("runner should not be called");
  });

  await assert.rejects(() => client.search(), CtxValidationError);
  await assert.rejects(() => client.search({ refresh: "off", limit: 5 }), CtxValidationError);
  await assert.rejects(() => client.search("   "), CtxValidationError);

  assert.equal(calls.length, 0);
});

test("wraps show and locate commands by ctx id and provider session id", async () => {
  const { client, calls } = mockClient(() => "{}");

  await client.showEvent("00000000-0000-0000-0000-000000000002", { window: 3 });
  await client.showSession("00000000-0000-0000-0000-000000000003", { mode: "full" });
  await client.showSession({ provider: "codex", providerSession: "codex-session", mode: "log" });
  await client.locateEvent("00000000-0000-0000-0000-000000000004");
  await client.locateSession({ provider: "codex", providerSession: "codex-session" });

  assert.deepEqual(
    calls.map((call) => call.args.slice(2)),
    [
      [
        "show",
        "event",
        "00000000-0000-0000-0000-000000000002",
        "--format",
        "json",
        "--window",
        "3",
      ],
      [
        "show",
        "session",
        "00000000-0000-0000-0000-000000000003",
        "--mode",
        "full",
        "--format",
        "json",
      ],
      [
        "show",
        "session",
        "--provider",
        "codex",
        "--provider-session",
        "codex-session",
        "--mode",
        "log",
        "--format",
        "json",
      ],
      ["locate", "event", "00000000-0000-0000-0000-000000000004", "--format", "json"],
      [
        "locate",
        "session",
        "--provider",
        "codex",
        "--provider-session",
        "codex-session",
        "--format",
        "json",
      ],
    ],
  );
});

test("reports versioning metadata", async () => {
  const { client } = mockClient(() => "ctx 1.2.3\n");

  assert.deepEqual(await client.version(), {
    schema_version: 1,
    api_version: AGENT_HISTORY_V1_VERSION,
    sdk_version: "0.0.0",
    adapter: "local-cli",
    ctx_version: "1.2.3",
  });
});

test("raises structured errors", async () => {
  const cli = createLocalAgentHistoryClient({
    runner: () => ({ exitCode: 2, stderr: "bad flag\n" }),
  });
  await assert.rejects(() => cli.status(), CtxCliError);

  const parse = createLocalAgentHistoryClient({ runner: () => "not json" });
  await assert.rejects(() => parse.status(), CtxParseError);

  await assert.rejects(() => parse.showEvent(""), CtxValidationError);
  await assert.rejects(() => parse.showSession({ provider: "codex" }), CtxValidationError);
});

test("raises timeout errors from the local adapter", async () => {
  const adapter = new (await import("../src/index.js")).LocalCliAdapter({
    ctxPath: process.execPath,
    timeoutMs: 1,
  });
  await assert.rejects(
    () => adapter.execute(["-e", "setTimeout(() => {}, 1000)"]),
    CtxTimeoutError,
  );
});

test("hosted client is an explicit placeholder", async () => {
  const client = createHostedAgentHistoryClient({ baseUrl: "https://ctx.example.invalid" });

  assert.equal((await client.version()).adapter, "hosted-placeholder");
  await assert.rejects(() => client.status(), CtxUnsupportedError);
});

test("dogfood toy app runs status/search/show/locate with mocked ctx", async () => {
  assert.deepEqual(await runDogfoodToy({ env: {} }), {
    ready: true,
    query: "local agent history",
    firstScope: "event",
    eventCount: 1,
    sessionMode: "lite",
    eventPath: "/tmp/ctx-sdk-dogfood/session.jsonl",
    sessionPath: "/tmp/ctx-sdk-dogfood/session.jsonl",
  });
});

test("shared agent-history-v1 fixtures use discriminated operation payloads", async () => {
  const fixturesDir = join(repoRoot, "contracts", "agent-history-v1", "fixtures");
  let entries = [];
  try {
    entries = await readdir(fixturesDir);
  } catch (error) {
    if (error.code !== "ENOENT") {
      throw error;
    }
  }

  const fixtureFiles = entries.filter((name) => name.endsWith(".json"));
  assert.notEqual(fixtureFiles.length, 0, "agent-history-v1 fixture directory should not be empty");
  for (const entry of fixtureFiles) {
    const fixture = JSON.parse(await readFile(join(fixturesDir, entry), "utf8"));
    const operation = operationFromFixtureName(entry);
    assert.equal(typeof fixture, "object", `${entry} should contain a JSON object`);
    assert.equal(fixture.contractVersion, AGENT_HISTORY_V1_VERSION, `${entry} contractVersion`);
    assert.equal(fixture.schemaVersion, 1, `${entry} schemaVersion`);
    assert.equal(fixture.operation, operation, `${entry} operation`);
    assertFixturePayload(entry, fixture);
  }
});

function operationFromFixtureName(name) {
  const operation = name.split(".")[0];
  switch (operation) {
    case "status":
    case "init":
    case "sources":
    case "import":
    case "sync":
    case "search":
    case "error":
      return operation;
    case "show-event":
      return "showEvent";
    case "show-session":
      return "showSession";
    case "locate-event":
      return "locateEvent";
    case "locate-session":
      return "locateSession";
    default:
      throw new Error(`unknown agent-history-v1 fixture operation in ${name}`);
  }
}

function assertFixturePayload(entry, fixture) {
  switch (fixture.operation) {
    case "status":
    case "init":
      assert.equal(typeof fixture.status.initialized, "boolean", `${entry} status.initialized`);
      assert.equal(typeof fixture.status.localOnly, "boolean", `${entry} status.localOnly`);
      break;
    case "sources":
      assert.ok(Array.isArray(fixture.sources), `${entry} sources`);
      assert.equal(typeof fixture.sources[0].provider, "string", `${entry} sources[0].provider`);
      assert.equal(typeof fixture.sources[0].importable, "boolean", `${entry} sources[0].importable`);
      break;
    case "import":
    case "sync":
      assert.equal(typeof fixture.import.resume, "boolean", `${entry} import.resume`);
      assert.equal(typeof fixture.import.totals, "object", `${entry} import.totals`);
      break;
    case "search":
      assert.ok(Array.isArray(fixture.search.results), `${entry} search.results`);
      if (fixture.search.results.length > 0) {
        assert.equal(
          typeof fixture.search.results[0].resultScope,
          "string",
          `${entry} search.results[0].resultScope`,
        );
      }
      break;
    case "showEvent":
      assert.ok(Array.isArray(fixture.event.events), `${entry} event.events`);
      assert.equal(typeof fixture.event.events[0].ctxEventId, "string", `${entry} event id`);
      break;
    case "showSession":
      assert.ok(Array.isArray(fixture.session.events), `${entry} session.events`);
      assert.equal(typeof fixture.session.mode, "string", `${entry} session.mode`);
      break;
    case "locateEvent":
    case "locateSession":
      assert.equal(typeof fixture.location.ctxSessionId, "string", `${entry} location session id`);
      assert.equal(typeof fixture.location.provider, "string", `${entry} location provider`);
      assert.equal(typeof fixture.location.source, "object", `${entry} location source`);
      break;
    case "error":
      assert.equal(typeof fixture.error.code, "string", `${entry} error.code`);
      assert.equal(typeof fixture.error.retryable, "boolean", `${entry} error.retryable`);
      break;
    default:
      throw new Error(`unsupported fixture operation ${fixture.operation} in ${entry}`);
  }
}

File: sdks/go/client_test.go
package ctxagenthistory

import (
	"context"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

func TestStatusDecodesAgentHistoryV1(t *testing.T) {
	client := NewClient(WithTransport(fakeTransport{
		response: `{
			"schema_version": 1,
			"initialized": true,
			"data_root": "/tmp/ctx",
			"database_path": "/tmp/ctx/history.sqlite3",
			"config_path": "/tmp/ctx/config.toml",
			"indexed_items": 7,
			"indexed_sources": 2,
			"cataloged_sessions": 3,
			"indexed_catalog_sessions": 2,
			"pending_catalog_sessions": 1,
			"failed_catalog_sessions": 0,
			"stale_catalog_sessions": 0,
			"local_only": true
		}`,
	}))

	status, err := client.Status(context.Background())
	if err != nil {
		t.Fatalf("Status returned error: %v", err)
	}
	if status.ContractVersion != APIVersion || status.Operation != "status" {
		t.Fatalf("unexpected envelope: %+v", status)
	}
	if !status.Status.Initialized || status.Status.IndexedItems != 7 || !status.Status.LocalOnly {
		t.Fatalf("unexpected status: %+v", status)
	}
}

func TestSearchBuildsAgentHistoryV1Operation(t *testing.T) {
	transport := &recordingTransport{response: `{
		"schema_version": 1,
		"query": "panic",
		"filters": {},
		"freshness": {"mode": "off", "status": "skipped", "source_count": 0, "totals": {}},
		"generated_at": "2026-01-01T00:00:00Z",
		"results": [],
		"pagination": {},
		"truncation": {}
	}`}
	client := NewClient(WithTransport(transport))
	semanticWeight := 0.35

	_, err := client.Search(context.Background(), SearchOptions{
		Query:                 "panic",
		Terms:                 []string{"sqlite", "retry"},
		Limit:                 5,
		Backend:               "hybrid",
		SemanticWeight:        &semanticWeight,
		Provider:              "codex",
		Workspace:             "ctx",
		Since:                 "30d",
		EventType:             "message",
		File:                  "crates/ctx-cli/src/main.rs",
		Session:               "00000000-0000-0000-0000-000000000001",
		Events:                true,
		Refresh:               "off",
		IncludeCurrentSession: true,
	})
	if err != nil {
		t.Fatalf("Search returned error: %v", err)
	}

	want := []string{
		"search", "panic", "--json", "--limit", "5",
		"--term", "sqlite", "--term", "retry",
		"--backend", "hybrid",
		"--semantic-weight", "0.35",
		"--provider", "codex",
		"--workspace", "ctx",
		"--since", "30d",
		"--event-type", "message",
		"--file", "crates/ctx-cli/src/main.rs",
		"--session", "00000000-0000-0000-0000-000000000001",
		"--refresh", "off",
		"--events",
		"--include-current-session",
	}
	if !reflect.DeepEqual(transport.op.Args, want) {
		t.Fatalf("args mismatch\nwant: %#v\n got: %#v", want, transport.op.Args)
	}
}

func TestSearchCamelizesRetrievalJSON(t *testing.T) {
	client := NewClient(WithTransport(fakeTransport{response: `{
		"schema_version": 1,
		"query": "agent history",
		"retrieval": {
			"requested_mode": "hybrid",
			"effective_mode": "lexical",
			"semantic_weight": 0.0,
			"semantic_fallback_code": "semantic_retrieval_failed",
			"semantic_fallback": "semantic_retrieval_failed",
			"coverage": {"embedded_items": 4, "indexed_now": 1},
			"diagnostics": {"query_embed_ms": 2}
		},
		"results": [{
			"result_scope": "event"
		}]
	}`}))

	response, err := client.Search(context.Background(), SearchOptions{Query: "agent history"})
	if err != nil {
		t.Fatalf("Search returned error: %v", err)
	}
	retrieval, ok := response.Search.Retrieval.(map[string]any)
	if !ok {
		t.Fatalf("top-level retrieval was not decoded: %#v", response.Search.Retrieval)
	}
	if retrieval["requestedMode"] != "hybrid" || retrieval["effectiveMode"] != "lexical" || retrieval["semanticWeight"] != 0.0 {
		t.Fatalf("top-level retrieval was not camelized: %#v", retrieval)
	}
	if retrieval["semanticFallbackCode"] != "semantic_retrieval_failed" {
		t.Fatalf("retrieval fallback code was not camelized: %#v", retrieval)
	}
	coverage, ok := retrieval["coverage"].(map[string]any)
	if !ok || coverage["embeddedItems"] != float64(4) || coverage["indexedNow"] != float64(1) {
		t.Fatalf("retrieval coverage was not camelized: %#v", retrieval)
	}
	diagnostics, ok := retrieval["diagnostics"].(map[string]any)
	if !ok || diagnostics["queryEmbedMs"] != float64(2) {
		t.Fatalf("retrieval diagnostics were not camelized: %#v", retrieval)
	}
}

func TestSearchRequiresQueryTermOrFileBeforeTransport(t *testing.T) {
	transport := &recordingTransport{response: `{"schema_version":1,"results":[]}`}
	client := NewClient(WithTransport(transport))

	for name, opts := range map[string]SearchOptions{
		"empty":        {},
		"filters only": {Refresh: "off", Limit: 5},
		"blank query":  {Query: "   "},
		"blank terms":  {Terms: []string{"", "   "}},
	} {
		t.Run(name, func(t *testing.T) {
			if _, err := client.Search(context.Background(), opts); !IsErrorKind(err, ErrorKindInvalidArgument) {
				t.Fatalf("Search error kind mismatch: %v", err)
			}
		})
	}
	if transport.op.Args != nil {
		t.Fatalf("Search invoked transport despite invalid input: %#v", transport.op.Args)
	}
}

func TestShowAndLocateValidateRequiredEventID(t *testing.T) {
	client := NewClient(WithTransport(fakeTransport{response: `{}`}))
	if _, err := client.ShowEvent(context.Background(), ShowEventOptions{}); !IsErrorKind(err, ErrorKindInvalidArgument) {
		t.Fatalf("ShowEvent error kind mismatch: %v", err)
	}
	if _, err := client.LocateEvent(context.Background(), LocateEventOptions{}); !IsErrorKind(err, ErrorKindInvalidArgument) {
		t.Fatalf("LocateEvent error kind mismatch: %v", err)
	}
}

func TestRejectsWrongCanonicalEnvelope(t *testing.T) {
	client := NewClient(WithTransport(fakeTransport{response: `{
		"contractVersion": "agent-history-v2",
		"schemaVersion": 1,
		"operation": "status",
		"backend": {"kind": "local"},
		"status": {"initialized": true, "localOnly": true}
	}`}))
	if _, err := client.Status(context.Background()); !IsErrorKind(err, ErrorKindUnsupportedSchema) {
		t.Fatalf("expected unsupported schema error, got %v", err)
	}

	client = NewClient(WithTransport(fakeTransport{response: `{
		"contractVersion": "agent-history-v1",
		"schemaVersion": 1,
		"operation": "search",
		"backend": {"kind": "local"},
		"status": {"initialized": true, "localOnly": true}
	}`}))
	if _, err := client.Status(context.Background()); !IsErrorKind(err, ErrorKindDecode) {
		t.Fatalf("expected operation decode error, got %v", err)
	}
}

func TestLegacyShowEventSourceObjectNormalizesToTypedEvent(t *testing.T) {
	client := NewClient(WithTransport(fakeTransport{response: `{
		"event": {
			"ctx_event_id": "event-1",
			"ctx_session_id": "session-1",
			"sequence": 1,
			"event_type": "message",
			"source": {
				"path": "/tmp/session.jsonl",
				"cursor": "line:1",
				"exists": true,
				"source_id": "source-1",
				"source_format": "codex_session_jsonl"
			},
			"text": "hello"
		},
		"events": []
	}`}))

	response, err := client.ShowEvent(context.Background(), ShowEventOptions{ID: "event-1"})
	if err != nil {
		t.Fatalf("ShowEvent returned error: %v", err)
	}
	if response.Event.Event == nil || response.Event.Event.Source != "" {
		t.Fatalf("unexpected normalized event source: %+v", response.Event.Event)
	}
	if response.Event.Source == nil || response.Event.Source.Path != "/tmp/session.jsonl" {
		t.Fatalf("expected source location from legacy event source, got %+v", response.Event.Source)
	}
}

func TestLocalCLIAdapterCommandFailureIsStructured(t *testing.T) {
	adapter := NewLocalCLIAdapter(WithCLIPath("ctx"))
	adapter.runner = fakeRunner{
		result: commandResult{
			Stderr:   []byte("no importable provider history sources found\n"),
			ExitCode: 1,
			Err:      errors.New("exit status 1"),
		},
	}

	_, err := adapter.Do(context.Background(), Operation{Name: "import", Args: []string{"import", "--json"}})
	var sdkErr *Error
	if !errors.As(err, &sdkErr) {
		t.Fatalf("expected structured error, got %T %v", err, err)
	}
	if sdkErr.Kind != ErrorKindCommandFailed || sdkErr.ExitCode != 1 || len(sdkErr.Command) != 3 {
		t.Fatalf("unexpected structured error: %+v", sdkErr)
	}
}

func TestLocalCLIAdapterClassifiesContextTimeout(t *testing.T) {
	adapter := NewLocalCLIAdapter(WithCLIPath("ctx"))
	adapter.runner = fakeRunner{result: commandResult{Err: context.DeadlineExceeded, ExitCode: -1}}

	_, err := adapter.Do(context.Background(), Operation{Name: "status", Args: []string{"status", "--json"}})
	if !IsErrorKind(err, ErrorKindTimeout) {
		t.Fatalf("expected timeout error, got %v", err)
	}
}

func TestLocalCLIAdapterAddsDataRootEnvironment(t *testing.T) {
	runner := &recordingRunner{result: commandResult{Stdout: []byte(`{"schema_version":1}`)}}
	adapter := NewLocalCLIAdapter(WithCLIPath("ctx"), WithDataRoot("/tmp/ctx-data"))
	adapter.runner = runner

	_, err := adapter.Do(context.Background(), Operation{Name: "status", Args: []string{"status", "--json"}})
	if err != nil {
		t.Fatalf("Do returned error: %v", err)
	}
	if !contains(runner.env, "CTX_DATA_ROOT=/tmp/ctx-data") {
		t.Fatalf("CTX_DATA_ROOT missing from env: %#v", runner.env)
	}
}

func TestHostedClientPlaceholder(t *testing.T) {
	client := NewHostedClient(HostedConfig{BaseURL: "https://example.invalid", APIKey: "test"})
	_, err := client.Status(context.Background())
	if !IsErrorKind(err, ErrorKindHostedNotImplemented) {
		t.Fatalf("unexpected hosted error: %v", err)
	}
	version, err := client.Version(context.Background())
	if err != nil {
		t.Fatalf("hosted Version returned error: %v", err)
	}
	if version.APIVersion != APIVersion || version.Transport != "hosted-placeholder" || version.CtxVersion != "" {
		t.Fatalf("unexpected hosted version: %+v", version)
	}
}

func TestVersionUsesTransport(t *testing.T) {
	client := NewClient(WithTransport(fakeTransport{response: "ctx 9.9.9\n"}))
	version, err := client.Version(context.Background())
	if err != nil {
		t.Fatalf("Version returned error: %v", err)
	}
	if version.APIVersion != APIVersion || version.SDKVersion != SDKVersion || version.CtxVersion != "ctx 9.9.9" {
		t.Fatalf("unexpected version: %+v", version)
	}
}

func TestContractErrorKindsArePublicConstants(t *testing.T) {
	for _, kind := range []ErrorKind{
		ErrorKindInvalidArgument,
		ErrorKindNotFound,
		ErrorKindNotInitialized,
		ErrorKindUnavailable,
		ErrorKindTimeout,
		ErrorKindCancelled,
		ErrorKindHostedNotImplemented,
		ErrorKindCommandFailed,
		ErrorKindDecode,
		ErrorKindUnknown,
	} {
		if kind == "" {
			t.Fatalf("empty error kind")
		}
	}
}

func TestCanonicalFixturesExposeTypedFields(t *testing.T) {
	search := readFixture[SearchResponse](t, "search.results.json")
	if search.ContractVersion != APIVersion || search.Operation != OperationSearch || search.Backend.Kind != BackendKindLocal {
		t.Fatalf("unexpected search envelope: %+v", search.Envelope)
	}
	if len(search.Search.Results) != 1 || search.Search.Results[0].WhyMatched[0] != "text" {
		t.Fatalf("unexpected typed search results: %+v", search.Search.Results)
	}
	if search.Search.Results[0].ResultType != "event" || search.Search.Results[0].Citations[0].TargetType != "event" {
		t.Fatalf("unexpected typed result/citation type: %+v", search.Search.Results[0])
	}
	if search.Search.Pagination == nil || search.Search.Pagination.Limit != 20 {
		t.Fatalf("unexpected pagination: %+v", search.Search.Pagination)
	}
	if search.Search.Truncation == nil || search.Search.Truncation.Truncated {
		t.Fatalf("unexpected truncation: %+v", search.Search.Truncation)
	}

	session := readFixture[ShowSessionResponse](t, "show-session.transcript.json")
	if session.Session.Session == nil || session.Session.Session.ProviderSessionID != "codex-fixture-session" {
		t.Fatalf("unexpected typed session: %+v", session.Session.Session)
	}

	location := readFixture[LocateEventResponse](t, "locate-event.location.json")
	if location.Location.Resume == nil || location.Location.Resume.Cursor != "line:2" {
		t.Fatalf("unexpected typed resume location: %+v", location.Location.Resume)
	}

	errorEnvelope := readFixture[ErrorResponse](t, "error.not-supported.json")
	if errorEnvelope.Error.Code != ErrorKindHostedNotImplemented || errorEnvelope.Backend.Kind != BackendKindHosted {
		t.Fatalf("unexpected error envelope: %+v", errorEnvelope)
	}
}

func TestContractFixturesIfPresent(t *testing.T) {
	fixtureRoot := filepath.Clean("../../contracts/agent-history-v1/fixtures")
	entries, err := os.ReadDir(fixtureRoot)
	if errors.Is(err, os.ErrNotExist) {
		t.Skip("agent-history-v1 fixtures are not present yet")
	}
	if err != nil {
		t.Fatalf("read fixture root: %v", err)
	}

	seen := false
	for _, entry := range entries {
		if entry.IsDir() || filepath.Ext(entry.Name()) != ".json" {
			continue
		}
		seen = true
		path := filepath.Join(fixtureRoot, entry.Name())
		data, err := os.ReadFile(path)
		if err != nil {
			t.Fatalf("read fixture %s: %v", path, err)
		}
		var envelope struct {
			Operation string          `json:"operation"`
			Response  json.RawMessage `json:"response"`
		}
		if err := json.Unmarshal(data, &envelope); err == nil && len(envelope.Response) > 0 {
			assertFixtureDecodes(t, path, envelope.Operation, envelope.Response)
			continue
		}
		assertFixtureDecodes(t, path, operationFromFilename(entry.Name()), data)
	}
	if !seen {
		t.Skip("agent-history-v1 fixture directory is present but empty")
	}
}

func assertFixtureDecodes(t *testing.T, path, operation string, data []byte) {
	t.Helper()
	var err error
	switch operation {
	case "status":
		var value StatusResponse
		err = json.Unmarshal(data, &value)
	case "init", "setup":
		var value InitResponse
		err = json.Unmarshal(data, &value)
	case "sources":
		var value SourcesResponse
		err = json.Unmarshal(data, &value)
	case "import", "sync":
		var value ImportResponse
		err = json.Unmarshal(data, &value)
	case "search":
		var value SearchResponse
		err = json.Unmarshal(data, &value)
	case "show_event", "showEvent":
		var value ShowEventResponse
		err = json.Unmarshal(data, &value)
	case "show_session", "showSession":
		var value ShowSessionResponse
		err = json.Unmarshal(data, &value)
	case "locate_event", "locateEvent":
		var value LocateEventResponse
		err = json.Unmarshal(data, &value)
	case "locate_session", "locateSession":
		var value LocateSessionResponse
		err = json.Unmarshal(data, &value)
	case "error":
		var value ErrorResponse
		err = json.Unmarshal(data, &value)
	default:
		var value map[string]any
		err = json.Unmarshal(data, &value)
	}
	if err != nil {
		t.Fatalf("decode fixture %s as %s: %v", path, operation, err)
	}
}

func readFixture[T any](t *testing.T, name string) T {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("../../contracts/agent-history-v1/fixtures", name))
	if errors.Is(err, os.ErrNotExist) {
		t.Skip("agent-history-v1 fixtures are not present yet")
	}
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	var value T
	if err := json.Unmarshal(data, &value); err != nil {
		t.Fatalf("decode fixture %s: %v", name, err)
	}
	return value
}

func operationFromFilename(name string) string {
	base := name[:len(name)-len(filepath.Ext(name))]
	if prefix, _, ok := strings.Cut(base, "."); ok {
		base = prefix
	}
	switch base {
	case "setup":
		return "init"
	case "show-event":
		return "showEvent"
	case "show-session":
		return "showSession"
	case "locate-event":
		return "locateEvent"
	case "locate-session":
		return "locateSession"
	default:
		return base
	}
}

type fakeTransport struct {
	response string
	err      error
}

func (f fakeTransport) Do(context.Context, Operation) ([]byte, error) {
	if f.err != nil {
		return nil, f.err
	}
	return []byte(f.response), nil
}

type recordingTransport struct {
	response string
	op       Operation
}

func (r *recordingTransport) Do(_ context.Context, op Operation) ([]byte, error) {
	r.op = op
	return []byte(r.response), nil
}

type fakeRunner struct {
	result commandResult
}

func (f fakeRunner) Run(context.Context, string, []string, []string) commandResult {
	return f.result
}

type recordingRunner struct {
	result commandResult
	path   string
	args   []string
	env    []string
}

func (r *recordingRunner) Run(_ context.Context, path string, args []string, env []string) commandResult {
	r.path = path
	r.args = append([]string(nil), args...)
	r.env = append([]string(nil), env...)
	return r.result
}

func contains(values []string, want string) bool {
	for _, value := range values {
		if value == want {
			return true
		}
	}
	return false
}
